File: android/app/src/main/java/com/navi/domain/models/Route.kt
package com.navi.domain.models

import java.util.Date

data class Route(
    val id: String,
    val origin: Coordinate,
    val destination: Coordinate,
    val waypoints: List<Coordinate> = emptyList(),
    val distance: Double, // meters
    val duration: Double, // seconds
    val trafficDuration: Double? = null, // seconds with traffic
    val polyline: String,
    val steps: List<RouteStep> = emptyList(),
    val alternatives: List<RouteAlternative>? = null,
    val createdAt: Date
)

data class RouteStep(
    val id: String,
    val instruction: String,
    val distance: Double,
    val duration: Double,
    val coordinate: Coordinate,
    val maneuver: Maneuver,
    val streetName: String? = null
)

data class Maneuver(
    val type: ManeuverType,
    val modifier: String? = null,
    val bearingBefore: Double? = null,
    val bearingAfter: Double? = null
)

enum class ManeuverType {
    TURN,
    NEW_NAME,
    DEPART,
    ARRIVE,
    MERGE,
    ON_RAMP,
    OFF_RAMP,
    FORK,
    END_OF_ROAD,
    CONTINUE,
    ROUNDABOUT,
    ROTARY,
    ROUNDABOUT_TURN,
    EXIT_ROUNDABOUT,
    EXIT_ROTARY
}

data class RouteAlternative(
    val id: String,
    val distance: Double,
    val duration: Double,
    val trafficDuration: Double? = null,
    val description: String,
    val polyline: String
)

data class SavedRoute(
    val id: String,
    val name: String,
    val route: Route,
    val userId: String,
    val notes: String? = null,
    val isFavorite: Boolean = false,
    val savedAt: Date
)

data class ActiveNavigation(
    val route: Route,
    val currentStepIndex: Int,
    val remainingDistance: Double,
    val remainingDuration: Double,
    val currentSpeed: Double? = null,
    val speedLimit: Double? = null,
    val nextManeuver: RouteStep? = null,
    val isOffRoute: Boolean = false
)

data class TrafficIncident(
    val id: String,
    val type: IncidentType,
    val severity: Severity,
    val description: String,
    val coordinate: Coordinate,
    val startTime: Date,
    val endTime: Date? = null,
    val affectedRoads: List<String> = emptyList()
)

enum class IncidentType {
    ACCIDENT,
    ROAD_CLOSURE,
    CONSTRUCTION,
    CONGESTION,
    WEATHER_HAZARD,
    POLICE_ACTIVITY,
    SPEED_CAMERA
}

enum class Severity {
    LOW,
    MEDIUM,
    HIGH,
    CRITICAL
}

data class RouteOptions(
    val avoidTolls: Boolean = false,
    val avoidHighways: Boolean = false,
    val avoidFerries: Boolean = false
)

File: ios/Navi/Models/Route.swift
import Foundation
import CoreLocation

struct Route: Codable, Identifiable {
    let id: String
    let origin: Coordinate
    let destination: Coordinate
    let waypoints: [Coordinate]
    let distance: Double // meters
    let duration: Double // seconds
    let trafficDuration: Double? // seconds with traffic
    let polyline: String
    let steps: [RouteStep]
    let alternatives: [RouteAlternative]?
    let createdAt: Date
}

struct RouteStep: Codable, Identifiable {
    let id: String
    let instruction: String
    let distance: Double
    let duration: Double
    let coordinate: Coordinate
    let maneuver: Maneuver
    let streetName: String?
}

struct Maneuver: Codable {
    let type: ManeuverType
    let modifier: String?
    let bearingBefore: Double?
    let bearingAfter: Double?
    
    enum ManeuverType: String, Codable {
        case turn
        case newName = "new name"
        case depart
        case arrive
        case merge
        case onRamp = "on ramp"
        case offRamp = "off ramp"
        case fork
        case endOfRoad = "end of road"
        case continue_
        case roundabout
        case rotary
        case roundaboutTurn = "roundabout turn"
        case exitRoundabout = "exit roundabout"
        case exitRotary = "exit rotary"
    }
}

struct RouteAlternative: Codable, Identifiable {
    let id: String
    let distance: Double
    let duration: Double
    let trafficDuration: Double?
    let description: String
    let polyline: String
}

struct SavedRoute: Codable, Identifiable {
    let id: String
    let name: String
    let route: Route
    let userId: String
    let notes: String?
    let isFavorite: Bool
    let savedAt: Date
}

struct ActiveNavigation: Codable {
    let route: Route
    let currentStepIndex: Int
    let remainingDistance: Double
    let remainingDuration: Double
    let currentSpeed: Double?
    let speedLimit: Double?
    let nextManeuver: RouteStep?
    let isOffRoute: Bool
}

struct TrafficIncident: Codable, Identifiable {
    let id: String
    let type: IncidentType
    let severity: Severity
    let description: String
    let coordinate: Coordinate
    let startTime: Date
    let endTime: Date?
    let affectedRoads: [String]
    
    enum IncidentType: String, Codable {
        case accident
        case roadClosure
        case construction
        case congestion
        case weatherHazard
        case policeActivity
        case speedCamera
    }
    
    enum Severity: String, Codable {
        case low
        case medium
        case high
        case critical
    }
}
